File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int zero=-1, ans=1, n=(int)nums.size();
        bool prob=false;
        for(int i=0;i<(int)nums.size();++i){
            if(nums[i]==0){
                if(zero==-1){
                    zero=i;
                }
                else{
                    prob=true;
                }
            }
            else ans*=nums[i];
        }
        vector<int> res(n);
        if(prob) {
            
        }
        else if(zero!=-1){
            res[zero]=ans;
        }
        else{
            for(int i=0;i<n;++i) res[i]=ans/nums[i];
        }
        return res;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n= nums.length;
        int[] ans= new int[n]; ans[0]= nums[0];
        for(int i=1;i<n;i++){
            ans[i]= ans[i-1]*nums[i];
        }
        int suffix= 1;
        for(int i=n-1;i>0;i--){
            ans[i]= ans[i-1]*suffix;
            suffix*= nums[i];
        }
        ans[0]= suffix;
        return ans;
    }
}